subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -0,0 +1,256 @@
+	}
+
+	return NULL;
+}
+
+static void select_next_image(void)
+{
+	ctx.cur_item_offset = 0;
+	ctx.cur_item_size = 0;
+
+	while (++ctx.cur_image_no < ctx.header.image_count && current_image_writer() == NULL) {
+		ctx.cur_offset += ctx.header.images[ctx.cur_image_no].size;
+	}
+
+	if (ctx.cur_image_no < ctx.header.image_count) {
+		ctx.cur_item_size = ctx.header.images[ctx.cur_image_no].size;
+	}
+}
+
+static int process_current_item(const uint8_t *chunk, size_t chunk_size)
+{
+	int err = 0;
+
+	/* Process only remaining bytes of the current item (header or image) */
+	chunk_size = MIN(chunk_size, ctx.cur_item_size - ctx.cur_item_offset);
+
+	if (ctx.cur_image_no < 0) {
+		memcpy(ctx.buffer + ctx.cur_item_offset, chunk, chunk_size);
+
+		if (ctx.cur_item_offset + chunk_size == ctx.cur_item_size) {
+			if (ctx.cur_image_no == IMAGE_NO_FIXED_HEADER) {
+				err = parse_fixed_header();
+			} else {
+				err = parse_cbor_header();
+			}
+		}
+	} else {
+		/* Image data */
+		const struct dfu_image_writer *writer = current_image_writer();
+
+		if (!writer) {
+			err = -ESPIPE;
+		}
+
+		if (!err && ctx.cur_item_offset == 0) {
+			err = writer->open(writer->image_id,
+					   ctx.header.images[ctx.cur_image_no].size);
+		}
+
+		if (!err) {
+			err = writer->write(chunk, chunk_size);
+		}
+
+		if (!err && ctx.cur_item_offset + chunk_size == ctx.cur_item_size) {
+			err = writer->close(true);
+		}
+	}
+
+	if (err) {
+		return err;
+	}
+
+	ctx.cur_offset += chunk_size;
+	ctx.cur_item_offset += chunk_size;
+
+	if (ctx.cur_item_offset == ctx.cur_item_size) {
+		select_next_image();
+	}
+
+	return chunk_size;
+}
+
+int dfu_multi_image_init(uint8_t *buffer, size_t buffer_size)
+{
+	if (buffer == NULL || buffer_size < FIXED_HEADER_SIZE) {
+		return -EINVAL;
+	}
+
+	memset(&ctx, 0, sizeof(ctx));
+	ctx.buffer = buffer;
+	ctx.buffer_size = buffer_size;
+	ctx.cur_image_no = IMAGE_NO_FIXED_HEADER;
+	ctx.cur_item_size = FIXED_HEADER_SIZE;
+
+	return 0;
+}
+
+int dfu_multi_image_register_writer(const struct dfu_image_writer *writer)
+{
+	if (ctx.writer_count == CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT) {
+		return -ENOMEM;
+	}
+
+	ctx.writers[ctx.writer_count++] = *writer;
+	return 0;
+}
+
+int dfu_multi_image_write(size_t offset, const uint8_t *chunk, size_t chunk_size)
+{
+	int result;
+	size_t chunk_offset = 0;
+
+	if (offset > ctx.cur_offset) {
+		/* Unexpected data gap */
+		return -ESPIPE;
+	}
+
+	while (1) {
+		/* Skip ahead to the current write offset */
+		chunk_offset += (ctx.cur_offset - offset);
+
+		if (chunk_offset >= chunk_size) {
+			break;
+		}
+
+		result = process_current_item(chunk + chunk_offset, chunk_size - chunk_offset);
+
+		if (result <= 0) {
+			return result;
+		}
+
+		chunk_offset += (size_t)result;
+		offset += (size_t)result;
+	}
+
+	return 0;
+}
+
+size_t dfu_multi_image_offset(void)
+{
+	return ctx.cur_offset;
+}
+
+int dfu_multi_image_done(bool success)
+{
+	const struct dfu_image_writer *writer = current_image_writer();
+	int err = 0;
```

Comment:
What do you think about resetting the `ctx.cur_offset` here?

Reply:
I don't have a strong opinion, but I thought that it would be reset on the next init, so we can keep the last used offset in case somebody wanted to use it for some progress bar etc.

Reply:
Agreed to leave as-is.